### PR TITLE
Support for Creating Git Version Tags (PR 1 of 2)

### PR DESCRIPTION
**Pull request by shundhammer**
## Trello

https://trello.com/c/9PGt7oCd/

## Bugzilla

https://bugzilla.suse.com/show_bug.cgi?id=1133435

## Problem

Our libyui community wants proper git tags for released versions of the libyui (libyui, libyui-qt, libyui-ncurses etc.) packages.

Use cases:

- Match bug fixes to the exact source trees used
- Proper dependencies between the various libyui packages with a well-defined source tree for each one

## Fix

Add some infrastructure for tagging source versions and use that when submitting source tarballs to OBS via Jenkins.

# Details

This PR includes a mix-in module `Packaging::GitHelpers` for various Git operations. This can be used in rake tasks. The code is in this package because this is shared between libyui and non-libyui YaST packages.

The most important method of this mix-in is `create_version_tag` which takes an optional code block to get the version number. The version number is in the .spec file for the great majority of our packages, but for libyui related packages, this is typically in a `VERSION.cmake` file in a format like this:

```cmake
SET( VERSION_MAJOR "3")
SET( VERSION_MINOR "5" )
SET( VERSION_PATCH "0" )
SET( VERSION "${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}${GIT_SHA1_VERSION}" )
```

This version number can be used in a rake task like this:

```rake
  desc "Create a git tag with this version"
  task :tag do
    create_version_tag { cmake_version }
  end
```

(where `cmake_version` is the parser function for this `VERSION.cmake` file that returns the version number as a string: "3.5.0")

To use the version from the (first) `package/*.spec` file, simply omit the code block; in that case, the `spec_version` function of the same mix-in module is used:

```rake
  desc "Create a git tag with this version"
  task :tag do
    create_version_tag
  end
```

This is mostly intended for non-libyui-related packages, but if everything is set up correctly and in sync, this also works just fine for libyui-related packages.

## Tag Names

- Master: Just the version number:  `3.5.0`
- Branch: Branch name and version number: `MyBranch-3.5.0`
- "Detached head" state: `detached-3.5.0`

If the tag already exists, the `create_version_tag` function just shows a message telling you that it exists (and returns _false_). For more fine control, use `tag_exists?`.

## Related PR

https://github.com/libyui/libyui-rake/pull/26

## Test

Manual tests with the changed libyui-rake (see the other PR) and local check-out directories of libyui packages. Used both `cmake_version` and `spec_version`. Tested with master, branches, detached head.

### CODE REVIEW

File: lib/packaging/git_helpers.rb [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    # Read the package version from the first .spec file in the packages/
+    # subdirectory.
+    #
+    # @return String
+    #
+    def spec_version
+      # use the first *.spec file found, assume all spec files
+      # contain the same version
+      File.readlines(Dir.glob("package/*.spec").first)
+          .grep(/^\s*Version:\s*/).first.sub("Version:", "").strip
```
Comment (jreidinger) [inline lib/packaging/git_helpers.rb]: :+1:

File: lib/packaging/git_helpers.rb [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    # For branches, this is "branchname-version".
+    # If in "detached head" state, this is "detached-version".
+    #
+    # Like in create_version_tag, the optional block can specify the version
+    # number to use. If no block is given, this uses spec_version, i.e. the
+    # version number of the first .spec file in the package/ subdirectory.
+    #
+    # @return String
+    #
+    def tag_name(&block)
```
Comment (jreidinger) [inline lib/packaging/git_helpers.rb]: just curious, why block instead of simple optional parameter like `def tag_name(version = spec_version)` ?

Comment (shundhammer) [reply lib/packaging/git_helpers.rb]: A moment a go I asked myself the same question. During developing this I rearranged the code several times; initially it had made more sense to do it like this.

And, of course, it looks a lot more Ruby'ish. :smile:

File: lib/packaging/git_helpers.rb [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  #
+  # Helpers around various git commands
+  module GitHelpers
+    # Create a git tag based on the version and the current branch.
+    #
+    # The optional block can specify the version number to use.  If no block
+    # is given, this uses spec_version, i.e. the version number of the first
+    # .spec file in the package/ subdirectory.
+    #
+    # @return String
```
[the remark this answers is not in the document]
Comment (shundhammer) [reply lib/packaging/git_helpers.rb]: Duh; that was a copy&paste error.

Actually, that string does not make any sense at all; but the caller might want to know if the tag was actually set or if it already existed. Adapting return value and docs.

File: lib/packaging/git_helpers.rb (hunk 1 above)
Comment (shundhammer) [reply lib/packaging/git_helpers.rb]: Stolen from existing code :smiley:

**Pull request merged**